### ISSUE TITLE
Problem while trying to sync a remote using Irmin_mirage

### ISSUE THREAD

**Issue by Engil** [mirage/irmin#334]
Hi!

My goal was to sync a remote repository with a local in-memory Irmin_git store, but I can't make it work within mirage.
I have the following config.ml

``` ocaml
open Mirage

let stack console = generic_stackv4 console tap0

let main =
  let libraries = ["irmin.git"; "mirage-http"; "irmin.mirage";"tls.mirage"] in
  let libraries = if get_mode () = `Xen then libraries else "irmin.unix" :: libraries in
  foreign
    ~libraries
    ~deps:[abstract nocrypto]
    ~packages:["irmin"; "mirage-http"; "mirage-flow";"tls";
               "mirage-types-lwt"; "channel"; "git"; "mirage-git"; "git-unix"]
    "Unikernel.Main" (console @-> resolver @-> conduit @-> job)

let () =
  let sv4 = stack default_console in
  let res_dns = resolver_dns sv4 in
  let conduit = conduit_direct sv4 ~tls:true in
  register "test" [
    main $ default_console $ res_dns $ conduit
  ]
```

and unikernel.ml

``` ocaml
open Lwt
open V1_LWT

module Main  (C: CONSOLE) (RES: Resolver_lwt.S) (CON: Conduit_mirage.S) = struct

  let task s = Irmin.Task.create ~date:0L ~owner:"Server" s

  let start console res ctx _ =

    let module Hash = Irmin.Hash.SHA1 in
    let module Context =
      ( struct
        let v _ = Lwt.return_some (res, ctx)
      end : Irmin_mirage.CONTEXT)
    in
    let module Mirage_git_memory = Irmin_mirage.Irmin_git.Memory(Context)(Git.Inflate.None) in
    let module Store = Mirage_git_memory(Irmin.Contents.String)(Irmin.Ref.String)(Hash) in
    let module Sync = Irmin.Sync(Store) in
    let config = Irmin_mem.config () in
    let new_task () = Store.Repo.create config >>= Store.master task in

    let uri = "git://github.com/Engil/KaamelottAsAService.git" in
    let upstream = Irmin.remote_uri uri in
    new_task () >>= fun t ->
    Lwt.catch
      (fun () -> Sync.pull_exn (t "Updating") upstream `Update)
      (fun e -> Lwt_io.printlf "%s: %s"
          uri (Printexc.to_string e)) >>= fun _ ->
    Lwt.return_unit

end


```

And after running mirage configure --unix --net socket, make, and then ./mir-test, I got the following error:

```
git://github.com/Engil/KaamelottAsAService.git: Invalid_argument("invalid bounds in Cstruct.get_uint8 [43,0](40356) off=0 len=1")
```

I don't know where I should start to understand what's going on, I tried with various repositories and the result is still the same. I am doing something wrong with Irmin's API ?

I can provide more details if needed.

Cheers,

**Comment by samoht** [comment]
I think it's because of this line:

```
let module Mirage_git_memory = Irmin_mirage.Irmin_git.Memory(Context)(Git.Inflate.None) in
```

Can you try with `Git_unix.Zlib` instead of `Git.Inflate.Context` and see if it works under unix? This was functorized if someone wants to change the compression algorithm with a different server implementation. Maybe not a good idea to expose that interface directly ...

**Comment by Engil** [comment]
Looks like it works using Git_unix.Zlib, thank you. :)
Is there a way to accomplish the same thing when not running under unix ?

**Comment by samoht** [comment]
You have two options to use that under Xen:
- you can either use the `irmin` branch on mirage-dev: https://github.com/mirage/mirage-dev/tree/irmin. It's not been updated since a long time so it's probably broken now but should be easy enough to upgrade. The branch contains a patched version of camlzip to run with Xen.

- or you can adapt [Decompress](https://github.com/oklm-wsh/Decompress) to work with Irmin. It's something I wanted to do for a long time and it's definitely the way forward, so if you have the time and patience, that'd be great if you can look at it and at least evaluate how hard it is to do the plumbing.

**Comment by Engil** [comment]
Ok,
Seems like @dinosaure is already working on making Decompress's interface similar to camlzip, so I guess I will wait for that before.
Thank you. :-)

**Comment by yomimono** [comment]
On 03/08/2016 03:08 PM, Enguerrand Decorne wrote:
>
> Ok,
> Seems like @dinosaure <https://github.com/dinosaure> is already 
> working on making Decompress's interface similar to camlzip, so I 
> guess I will wait for that before.
> Thank you. :-)
>

You may be interested in some work I did getting a MirageOS unikernel to 
speak with GitHub and a remote Irmin server, which uses Decompress:

https://github.com/yomimono/example-unikernels/blob/store_in_irmin/https-client/unikernel.ml

Thanks,
Mindy

**Comment by hannesm** [comment]
I think this is fixed with decompress 0.4...

**Comment by samoht** [comment]
I am closing the issue seems it seems to be fixed upstream. I really want to have Decompress be the default for `ocaml-git`, but this needs more work. Luckily @dinosaure is working on this!
